### PR TITLE
[Bug]Batch: persist runtime ownership to recover safely during MDS rolling updates

### PR DESCRIPTION
**Pull request by zhangjyr**
## Pull Request Description
During an MDS rolling update, a new MDS can come up before the old MDS has fully released or torn down the runtime it was managing. In that window, the replacement instance may see an existing runtime but cannot safely assume whether it is reusable, still owned by a live worker, or already being deleted.

This can lead to duplicate provisioning, unsafe teardown, lost ownership, or jobs getting stuck during recovery.

### Runtime recovery with persisted ownership
- persist runtime ownership metadata so recovery decisions survive MDS restarts and rolling updates
- track owner identity, worker ownership, heartbeat, and delete lifecycle on the runtime reference
- treat reconnect/load as a first-class recovery path instead of always provisioning a new runtime
- detect when a runtime is still owned by another live worker and retry instead of stealing or tearing it down
- detect delete-in-progress state and wait for teardown completion before reprovisioning
- refresh heartbeat during liveness checks so ownership can expire safely when the original owner is gone

### Safer batch recovery flow
- allow recovered jobs to resume correctly when MDS restarts in the middle of validation, execution, or finalization
- resume partial execution from the next unfinished request instead of replaying the whole batch
- support finalize-only recovery paths when execution is already complete
- improve orphaned execution cleanup for jobs recovered after restart or expiry

### E2E abnormal test hardening
- expand abnormal end-to-end coverage for crash, restart, ownership handoff, teardown, and finalize-edge cases
- cover the rolling-update scenario where the new MDS starts while the old MDS still owns or is deleting the runtime
- refactor the abnormal test harness so the same recovery scenarios can be exercised across multiple runtime backends
- add reusable backend helpers to make future backend coverage easier to extend

## Why
The key goal is to make rolling updates safe. A new MDS should be able to recover a job and its runtime without guessing. Persisted runtime ownership gives recovery logic enough information to decide whether to reconnect, wait, retry, or provision fresh infrastructure.

This reduces the risk of:
- duplicate runtime creation
- teardown races between old and new MDS instances
- reclaiming a runtime that is still live elsewhere
- batches getting stranded during crash or rollout recovery

**Important: Before submitting, please complete the description above and review the checklist below.**

---

<details>
<summary><strong>Contribution Guidelines (Expand for Details)</strong></summary>

<p>We appreciate your contribution to aibrix! To ensure a smooth review process and maintain high code quality, please adhere to the following guidelines:</p>

<h3>Pull Request Title Format</h3>
<p>Your PR title should start with one of these prefixes to indicate the nature of the change:</p>
<ul>
    <li><code>[Bug]</code>: Corrections to existing functionality</li>
    <li><code>[CI]</code>: Changes to build process or CI pipeline</li>
    <li><code>[Docs]</code>: Updates or additions to documentation</li>
    <li><code>[API]</code>: Modifications to aibrix's API or interface</li>
    <li><code>[CLI]</code>: Changes or additions to the Command Line Interface</li>
    <li><code>[Misc]</code>: For changes not covered above (use sparingly)</li>
</ul>
<p><em>Note: For changes spanning multiple categories, use multiple prefixes in order of importance.</em></p>

<h3>Submission Checklist</h3>
<ul>
    <li>[ ] PR title includes appropriate prefix(es)</li>
    <li>[ ] Changes are clearly explained in the PR description</li>
    <li>[ ] New and existing tests pass successfully</li>
    <li>[ ] Code adheres to project style and best practices</li>
    <li>[ ] Documentation updated to reflect changes (if applicable)</li>
    <li>[ ] Thorough testing completed, no regressions introduced</li>
</ul>

<p>By submitting this PR, you confirm that you've read these guidelines and your changes align with the project's contribution standards.</p>

</details>

### CODE REVIEW

File: python/aibrix/aibrix/batch/job_driver/runtime/base.py [hunk 1]
```diff
@@ -367,6 +492,258 @@ def _build_runtime_ref(
     def _load_runtime_ref(self, job: BatchJob) -> Optional[JobRuntimeRef]:
         return job.status.get_runtime_ref(self._get_runtime_key(job))
 
+    def _runtime_ref_lease_seconds(self) -> float:
+        if self.session_liveness_check_interval_s > 0:
+            return max(self.session_liveness_check_interval_s * 2, 1.0)
```
Comment (DwyaneShi) [inline python/aibrix/aibrix/batch/job_driver/runtime/base.py:497]: let's use 30.0 instead of 1.0 to avoid nonsense small values

File: python/aibrix/aibrix/batch/job_driver/runtime/base.py [hunk 2]
```diff
@@ -237,6 +326,42 @@ async def _check_liveness(self, handle: Any) -> None:
...
+        self,
+        handle: Any,
+        *,
+        job: Optional[BatchJob] = None,
+        progress_manager: Optional[RunningJobs] = None,
+        worker_id: Optional[str] = None,
+    ) -> bool:
+        """Poll until runtime confirm to be deleted"""
+        loop = asyncio.get_running_loop()
+        deadline = loop.time() + float(self._ready_timeout_seconds)
```
Comment (DwyaneShi) [inline python/aibrix/aibrix/batch/job_driver/runtime/base.py:339]: why using _ready_timeout_seconds? should we have a dedicated timeout?

File: python/aibrix/aibrix/batch/job_driver/runtime/base.py [hunk 3]
```diff
@@ -544,9 +962,72 @@ def _should_teardown_failed_wait_ready(self, exc: Exception) -> bool:
...
+        try:
+            while True:
+                await asyncio.sleep(self.session_liveness_check_interval_s)
+                if self._stop_requested.is_set():
+                    return
+
+                try:
+                    await self._check_liveness(handle)
+                    consecutive_failures = 0
+                except BaseException as exc:
```
Comment (DwyaneShi) [inline python/aibrix/aibrix/batch/job_driver/runtime/base.py:999]: Is BaseException able to cover all exceptions? better to use Exception if we're not 100% confident

File: python/aibrix/aibrix/batch/job_driver/runtime/base.py (hunk 1 above)
Comment (zhangjyr) [reply python/aibrix/aibrix/batch/job_driver/runtime/base.py:497]: The session_liveness_check_interval_s is set to 30s; 1.0 is here for the sanity check (though it seems unnecessary)

File: python/aibrix/aibrix/batch/job_driver/runtime/base.py (hunk 2 above)
Comment (zhangjyr) [reply python/aibrix/aibrix/batch/job_driver/runtime/base.py:339]: I hope to eliminate not-must configurations; the tear-down timeout is set to _ready_timeout_seconds, assuming that runtime start-up and tear-down time costs are symmetrical, or tear-down is much quicker than start-up. Since this value can be passed from the batch UI, I prefer to use it instead of a fixed tear-down timeout.

File: python/aibrix/aibrix/batch/job_driver/runtime/base.py (hunk 3 above)
Comment (zhangjyr) [reply python/aibrix/aibrix/batch/job_driver/runtime/base.py:999]: Yes, BaseException can cover asyncio.CancelledError that the Exception can not cover.
